internal/search/job/jobutil/feeling_lucky_search_job.go

```diff
@@ -0,0 +1,120 @@
+package jobutil
+
+import (
+	"github.com/sourcegraph/sourcegraph/internal/search/job"
+	"github.com/sourcegraph/sourcegraph/internal/search/query"
+	"github.com/sourcegraph/sourcegraph/internal/search/run"
+)
+
+// NewFeelingLuckySearchJob generates opportunistic search queries by applying
+// various rules in sequence, transforming the original input plan into various
+// queries that alter its interpretation (e.g., search literally for quotes or
+// not, attempt to search the pattern as a regexp, and so on).
+//
+// Generated queries return a resulting list of jobs. The application order of
+// rules is deterministic. This means the query order, and therefore runtime
+// execution, outputs the same search results for the same inputs. I.e., there
+// is no random choice when applying a rule.
+func NewFeelingLuckySearchJob(inputs *run.SearchInputs, plan query.Plan) []job.Job {
+	children := make([]job.Job, 0, len(plan))
+	for _, b := range plan {
+		for _, newBasic := range applyRulesList(b, rulesList...) {
+			child, err := NewBasicJob(inputs, newBasic)
+			if err != nil {
+				return nil
+			}
+			children = append(children, child)
+		}
+	}
+	return children
+}
+
+var rulesList = [][]rule{
+	{unquotePatterns},
+}
+
+// rule represents a transformation function on a Basic query. Applying rules
+// cannot fail: either they apply and produce a valid, non-nil, Basic query, or
+// they cannot apply, in which case they return nil. See the `unquotePatterns`
+// rule for an example.
+type rule func(query.Basic) *query.Basic
+
+// applyRulesList takes a list of lists of rules. The order of rules in the inner
+// lists represent rule composition. Each list of rules in the outer list
+// represent one possible query production, if the sequence of the rules in this
+// list apply successfully. Example:
+//
+// If we have input rule list  [ [ R1, R2 ], [ R2 ] ] and input query B0, then:
+//
+// - If both inner lists apply, we get an output [ B1, B2] where B1 is generated
+//   from applying R1 then R2, and B2 is generated from just applying R2.
+// - If only the first inner list applies, R1 then R2, we get the output [ B1 ]
+// - If only the second inner list applies, R2 on its own, we get the output [ B2 ]
+func applyRulesList(b query.Basic, rulesList ...[]rule) []query.Basic {
+	bs := []query.Basic{}
+	for _, l := range rulesList {
+		if generated := applyRules(b, l...); generated != nil {
+			bs = append(bs, *generated)
+		}
+	}
+	return bs
+}
+
+// applyRules applies every rule in sequence to `b`. If any rule does not apply, it returns nil.
+func applyRules(b query.Basic, rules ...rule) *query.Basic {
+	if len(rules) == 0 {
+		return &b
+	}
+	if next := rules[0](b); next != nil {
+		return applyRules(*next, rules[1:]...)
+	}
+	return nil
+}
```

Comment:
nit: iteration rather than recursion? I find this a bit easier to read.

```suggestion
func applyRules(b query.Basic, rules ...rule) *query.Basic {
	for _, rule := range rules {
		b = rule(b)
		if b == nil {
			return nil
		}
	}
	return b
}
```

Reply:
Doesn't type check because pointers 😄 This is why I went for recursion, my brain starts breaking when I have to keep track of intermediate state to overcome type checking, versus just having the logic drive intermediate state. I think this is the best imperative version I can come up with:

```go
func applyRules(b query.Basic, rules ...rule) *query.Basic {                                                                                                                                                                           
        if len(rules) == 0 {                                                                                                                                                                                                           
                return nil                                                                                                                                                                                                             
        }                                                                                                                                                                                                                              
        current := rules[0](b)                                                                                                                                                                                                         
        if current == nil {                                                                                                                                                                                                            
                return nil                                                                                                                                                                                                             
        }                                                                                                                                                                                                                              
                                                                                                                                                                                                                                       
        for _, rule := range rules[1:] {                                                                                                                                                                                               
                current = rule(*current)                                                                                                                                                                                               
                if current == nil {                                                                                                                                                                                                    
                        return nil                                                                                                                                                                                                     
                }                                                                                                                                                                                                                      
        }                                                                                                                                                                                                                              
        return current                                                                                                                                                                                                                 
}          
```

More "Go readable" but more "fussy" to me (acknowledging recursion is much more natural to me than usual Go coders). If you can come up with a better imperative version lmk

Reply:
Oh whoops. My bad for suggesting code I didn't try to compile. 

What about this? It does actually compile 😄 

```go
// applyRules applies every rule in sequence to `b`. If any rule does not apply, it returns nil.
func applyRules(b query.Basic, rules ...rule) *query.Basic {
	for _, rule := range rules {
		res := rule(b)
		if res == nil {
			return nil
		}
		b = *res
	}
	return &b
}
```

Also though, feel free to ignore because this is 100% nit and is just asking you to cater to my imperative lifestyle.

Reply:
works for me. I don't mind catering as long as I don't have to deal with actually finding a succinct and correct imperative style :-) just has a tendency to break my flow